pkg/ccl/backupccl/backup_metadata.go

```diff
@@ -0,0 +1,1135 @@
+	var desc *descpb.Descriptor
+	if len(wrapper.value) > 0 {
+		desc = &descpb.Descriptor{}
+		err := protoutil.Unmarshal(wrapper.value, desc)
+		if err != nil {
+			return err
+		}
+	}
+
+	id, err := decodeDescSSTKey(wrapper.key.Key)
+	if err != nil {
+		return err
+	}
+
+	*rev = BackupManifest_DescriptorRevision{
+		Desc: desc,
+		ID:   id,
+		Time: wrapper.key.Timestamp,
+	}
+	return nil
+}
+
+// StatsIterator is a simple iterator to iterate over stats.TableStatisticProtos.
+type StatsIterator struct {
+	backing bytesIter
+	err     error
+}
+
+// StatsIter creates a new StatsIterator for the backup metadata.
+func (b *BackupMetadata) StatsIter(ctx context.Context) StatsIterator {
+	backing := makeBytesIter(ctx, b.store, b.filename, []byte(sstStatsPrefix), b.enc, false)
+	return StatsIterator{
+		backing: backing,
+	}
+}
+
+// Close closes the iterator.
+func (si *StatsIterator) Close() {
+	si.backing.close()
+}
+
+// Err returns the iterator's error.
+func (si *StatsIterator) Err() error {
+	if si.err != nil {
+		return si.err
+	}
+	return si.backing.err()
+}
+
+// Next retrieves the next stats proto in the iterator.
+//
+// Next returns true if next element was successfully unmarshalled into
+// statsPtr, and false if there are no more elements or if an error was
+// encountered. When Next returns false, the user should call the Err method to verify the
+// existence of an error.
+func (si *StatsIterator) Next(statsPtr **stats.TableStatisticProto) bool {
+	wrapper := resultWrapper{}
+	ok := si.backing.next(&wrapper)
+
+	if !ok {
+		return false
+	}
+
+	var s stats.TableStatisticProto
+	err := protoutil.Unmarshal(wrapper.value, &s)
+	if err != nil {
+		si.err = err
+		return false
+	}
+
+	*statsPtr = &s
+	return true
+}
+
+type bytesIter struct {
+	Iter storage.SimpleMVCCIterator
+
+	prefix      []byte
+	useMVCCNext bool
+	iterError   error
+}
+
+func makeBytesIter(
+	ctx context.Context,
+	store cloud.ExternalStorage,
+	path string,
+	prefix []byte,
+	enc *jobspb.BackupEncryptionOptions,
+	useMVCCNext bool,
+) bytesIter {
+	var encOpts *roachpb.FileEncryptionOptions
+	if enc != nil {
+		key, err := getEncryptionKey(ctx, enc, store.Settings(), store.ExternalIOConf())
+		if err != nil {
+			return bytesIter{iterError: err}
+		}
+		encOpts = &roachpb.FileEncryptionOptions{Key: key}
+	}
+
+	iter, err := storageccl.ExternalSSTReader(ctx, store, path, encOpts)
+	if err != nil {
+		return bytesIter{iterError: err}
+	}
+
+	iter.SeekGE(storage.MakeMVCCMetadataKey(prefix))
+	return bytesIter{
+		Iter:        iter,
+		prefix:      prefix,
+		useMVCCNext: useMVCCNext,
+	}
+}
+
+func (bi *bytesIter) next(result interface{}) bool {
+	if bi.iterError != nil {
+		return false
+	}
+
+	valid, err := bi.Iter.Valid()
+	if err != nil || !valid || !bytes.HasPrefix(bi.Iter.UnsafeKey().Key, bi.prefix) {
+		bi.close()
+		bi.iterError = err
+		return false
+	}
+
+	resWrapper, ok := result.(*resultWrapper)
```

Comment:
huh, do we ever pass `next` something that _isn't_ a *resultWrapper? if no, could we just make that its arg type?

Reply:
Done, made *resultWrapper the arg type.